calyx-py/calyx/queues.py

```diff
@@ -192,6 +192,109 @@ def __len__(self) -> int:
         return self.len
 
 
+@dataclass
+class RRQueue:
+   """
+   This is a version of a PIFO generalized to n flows, with a round robin policy
+   that is work conserving. Just as with the PIFO, if a flow falls silent, the 
+   remaining flows will take their turns. That flow effectively skips its turn.
+
+   Supports the operations `push`, `pop`, and `peek`.
+   It takes in a list `boundaries` that must be of length `n`, the user can
+   divide the flows however they like.
+
+   - Push works analagously to that of the PIFO, except it checks the `boundaries`
+   list to determine which flow to push to.
+   - Pop first tries to pop from `hot`, given the length is not 0. If this fails, 
+   it continues to check all flows after. It then increments hot at the end and
+   decreases the length.
+   - Peek works the same as `pop`, except `hot` remains unchanged.
+   """
+
+   def __init__(self, n, boundaries, max_len: int):
+       self.data = []
+       self.hot = 0
+       self.n_flows = n
+       self.pifo_len = 0
+       self.boundaries = boundaries
+       for i in range(n):
+           queue = Fifo(max_len)
+           self.data.append(queue)
+
+       self.max_len = max_len
+       assert (
+           self.pifo_len <= self.max_len
+       )  # We can't be initialized with a PIFO that is too long.
+
+
+   def push(self, val: int):
+       """Pushes `val` to the PIFO."""
+       if self.pifo_len == self.max_len:
+            raise QueueError("Cannot push to full PIFO.")
+       for b in range(len(self.boundaries)):
+          if val <= self.boundaries[b]:
+            self.data[b].push(val)
+            self.pifo_len += 1
+            break
+
+
+   def increment_hot(self):
+       """Increments hot, taking into account wrap around."""
+       if self.hot == (self.n_flows - 1): # handle wrap around when updating hot
+            self.hot = 0
+       else:
+            self.hot = self.hot + 1
+
+        
+   def pop(self) -> Optional[int]:
+       """Pops the PIFO. Updates `hot`."""
+       if self.pifo_len == 0:
+        raise QueueError("Cannot pop from empty PIFO.")
+       
+       original_hot = self.hot
+
+       while True:
+        try:
+            val = self.data[self.hot].pop()
+            if val is not None:
+                self.increment_hot() 
+                self.pifo_len -= 1              
+                return val
+            else:
+                self.increment_hot()
+                if self.hot == original_hot:
```

Comment:
So now that things work, would you like to try and take out this test and see if everything still works?